### ISSUE TITLE
Reopeinig: thread-safety fail: clblasDgemm assertion fail within openmp-environment using multiple GPUs

### ISSUE THREAD

**Issue by paolodalberto** [clMathLibraries/clBLAS#269]
The code follows a multithreaded program calling 2 gemm in parallel. The code has been made thread safe 

An example using clBLAS-2.11.0-Linux-x64 (package from master)
./sgemm2 2 10000 10000 10000 10 0
----------> get time 4.755824e+01 sec<------
2 GFLOPS 420.537027

the code on the release: 
using directly the Fiji Release clBLAS-2.10.0-Fiji-Linux-x64-CL2.0
OpenCL error -45 on line 281 of /home/fpadmin/Timmy/clBLAS2-10/clBLAS/src/library/blas/xgemm.cc
Segmentation fault (core dumped)

but considering the previous performance I do not think is a fix 
With the old /opt/clAmdBlas-1.10.321
----------> get time 8.035512e+00 sec<------
2 GFLOPS 2,488.951544

**Comment by paolodalberto** [comment]
The single threaded is very promising that why I Am asking for the fix :
 ./sgemm 1 10000 10000 10000 10 0
N of GPUS 3 
Start 
----------> get time 6.613615e+00 sec<------
1 GFLOPS 3024.064751 
done

faster than 2 GPUs using the previous old kernels ....

**Comment by paolodalberto** [comment]
Well I can always make the calls of gems without blocking ... let me check if this will work

**Comment by paolodalberto** [comment]
./sgemm3 1 15000 15000 15000 10 0
----------> get time 2.334493e+01 sec<------
1 GFLOPS 2891.419488 
making everything asynchronous (non blocking) seems helping but still a far cry 

If I run two regular sgemm using different GPU but different program

./sgemm 1 15000 15000 15000 10 0 & ./sgemm 2 15000 15000 15000 10 0
[1] 8753
N of GPUS 3 
N of GPUS 3 
Start 
Start 
----------> get time 2.236617e+01 sec<------
1 GFLOPS 3017.950636 
done
----------> get time 2.266387e+01 sec<------
2 GFLOPS 2978.309304 
done

I know I should be able to achieve about 6TFLOPS .... 
[multigraphic_sgemm.zip](https://github.com/clMathLibraries/clBLAS/files/273895/multigraphic_sgemm.zip)

**Comment by kknox** [comment]
Hi @paolodalberto 
Can you confirm my understanding?
1.  You are comparing /develop branch vs /master branch
2.  /master branch as a threading problem, but /develop has the threading fix
3.  /master branch as better performance?  in /develop or /master?
4.  What is the difference between sgemm, sgemm2 and sgemm3?  I don't understand

Try measuring the performance of 15072; ideally, you would like to keep m,n,k as multiples of 96 if you can.

**Comment by paolodalberto** [comment]
Hi @kknox Kent,

1. You are comparing /develop branch vs /master branch
I create my package from master ( I do not know which branch has the fix) 

2. /master branch as a threading problem, but /develop has the threading fix
The master has a thread fix as well, but the parallelism is lost. That is, it is possible to run multithreaded gemm without any error (while generating the kernels). The fiji release has problem using multithreaded gemms.

3. /master branch as better performance? in /develop or /master?
I did not build from develop 

4. What is the difference between sgemm, sgemm2 and sgemm3? I don't understand
I appended the code for sgemm2 and sgemm3. 
a) sgemm is an example where sgemm is called where I can specify the GPU I want to run (0,1,2)
b) sgemm2 is a multithreaded example where I call two sgemm: on GPU 1 and 2 (fiji Pro-Duo)
c) sgemm3 is a single thread but calling two sgemm using non blocking methods. 

The test will be on a range and I will try on a 2xHawaii system ... (and if I can 1hawaii and 2 Fijis). I am not looking for the upper bound (certainly I could use x96 matrices ....) I am trying to understand what is the performance gap between 1 and 2 GPUs ... 

Can you create a release package for Fiji and Hawaii? with the thread bug fix and kernel precompiled ... for you is matter of few minutes for me I am just started in using cmake and its magic 
Thanks
Paolo

**Comment by pavanky** [comment]
@paolodalberto

 `clWaitForEvents` are blocking calls.
[Reference](https://www.khronos.org/registry/cl/sdk/1.0/docs/man/xhtml/clWaitForEvents.html):
> Waits on the host thread for commands identified by event objects in event_list to complete. A command is considered complete if its execution status is CL_COMPLETE or a negative value.

This is the reason you are seeing `sgemm3` as having only half the performance.


As for multi threaded performance, you have `pthread_create` inside the timer.. Have you seen how much time `pthread_create` takes without calling clBlas ?

**Comment by paolodalberto** [comment]
1) I stand corrected I am building from develop branch
git clone --depth=50 --branch=develop https://github.com/clMathLibraries/clBLAS.git

2) I did not check how expensive is creating the thread, but it is a good point. In my previous ( 1 year ago) I got speed up using thread so I did not consider the problem the thread creation ...

3) yep, I am working on putting the events into the appropriate queues .. I will show

**Comment by paolodalberto** [comment]
Now I understand why this did not go into the release. 

cgemm will not work using the develop branch, it does work with Fiji Release. 

Very likely I will be working without the threads ....

**Comment by kknox** [comment]
Why does cgemm not work using the /develop branch?  It should.

**Comment by paolodalberto** [comment]
segmentation fault in  the kernel (Hawaii release as well)
how would you test cgemm ? 

Hawaii/Hawaii-cgemm 1 100 100 100 10
Start 
Segmentation fault (core dumped)

No problem with the Fiji release on a Fiji board.

template for all types
[sgemm.template..zip](https://github.com/clMathLibraries/clBLAS/files/276618/sgemm.template.zip)
My make file
[Makefile.zip](https://github.com/clMathLibraries/clBLAS/files/276619/Makefile.zip)

Same problem for the develop branch but not for the Fiji Release ....

I will share the performance for ProDuo (Fiji) and 295x2 (Hawaii)  .. 
Fiji hands down faster for single precision 
Hawaii hands down faster for double precision


Question how would you make it work if I have a fiji and an hawaii GPUs ? is it possible ?

**Comment by paolodalberto** [comment]
Using the Hawaii release I could not run any test. So I run on 295x2 using the Fiji Release. 

[performanceinGFLOPS_fiji_hawaii.zip](https://github.com/clMathLibraries/clBLAS/files/276649/performanceinGFLOPS_fiji_hawaii.zip)

multiple GPUs are run without threads (using non blocking) 

You will notice that there are missing points for the problem size 8,000, single precision for Fiji and Hawaii for the non blocking: segmentation fault (same thing for double precision size 4000).

Let me know if these problems, like seg. fault, it is an ubuntu issue or you can reproduce it in your environment. Also You may notice that Hawaii kernels for double are far superior .... 

I will try to run 3 way (... and then come back to you

**Comment by paolodalberto** [comment]
Nice I can see all 4 GPUs ....

**Comment by paolodalberto** [comment]
some news .... I upgrade my PSUs so I can work with one 295x and ProDuo (the 295x is on the first slot and thus I cannot use all its potential I will work on that. But a first round of performance tests:

[Results.zip](https://github.com/clMathLibraries/clBLAS/files/287161/Results.zip)

These results are computed using the non blocking strategy.

Question: 
typedef std::map<kernel_map_key, cl_kernel> kernel_map_t;    
This is used in the xgemm and I think this is troublesome with thread ... cannot be initialized at clBLAS initialization ? ( err = clAmdBlasSetup()) reading it should be safe .... and pre-compiling the kernels will be good ...

**Comment by paolodalberto** [comment]
FYI:  I cannot pre-compile the kernels ...

**Comment by paolodalberto** [comment]
A note about the failure for the problem 8000x8000x8000

the log 
sgemm_Col_NN_B1_4000x8000x8000
For M,N,K = 8000,4000,8000 and 44 CUs selected tile is wg=16x16, microTile=5x5, macroTile=80x80 kernelsNeeded=1,0,0,0
makeGemmKernel: 30446480 Creating program from source 
makeGemmKernel: 30446480 Created program from source (null) 
makeGemmKernel now built; returning.
sgemm_Col_NN_B1_4000x8000x8000
For M,N,K = 8000,4000,8000 and 64 CUs selected tile is wg=16x16, microTile=5x5, macroTile=80x80 kernelsNeeded=1,0,0,0
makeGemmKernel: 31143872 Creating program from source 
makeGemmKernel: 31143872 Created program from source (null) 
Segmentation fault (core dumped)

it fails even for two of the same GPUs

**Comment by paolodalberto** [comment]
What I mean is that the compilation of the kernel fails, the second time around for that specific problem size @kknox 

How can I help you debugging this ?

**Comment by kknox** [comment]
Hi @paolodalberto 
I think this thread got to the point where it was tracking multiple issues; lets bring this back to just one.  

1.  You are having kernels fail to compile (only on second run?), and your repro case is posted to [this message](https://github.com/clMathLibraries/clBLAS/issues/269#issuecomment-220861745)?  
2.  This problem only reproduces on Hawaii hardware?  Works on fiji?
3.  Are you using the /develop branch or are you using a release package?

In other words, can you concisely state everything I need to reproduce the failure?

**Comment by paolodalberto** [comment]
On Hawaii release, Fiji Release, and develop: they all have problem with that particular problem size.  In the message above I activate the debug flag in the develop branch so to highlight where the problem happens. For the Release, of  course I do not know if it is the same problem (compilation) but it happens for the same problem size and same parallel computation across all configurations.  

The test that fails is not multithreaded, it is non blocking. However, I believe the compilation is sequential. 

I will provide my code and my tests, so you may try to reproduce the case using multiple GPUs, next posts.

Thank you

**Comment by paolodalberto** [comment]
let me attach here the tests, the code and the results for your convenience

[test.clblas.06.02.2015.zip](https://github.com/clMathLibraries/clBLAS/files/296062/test.clblas.06.02.2015.zip)

This is the summary of the report in a more formal venue
http://www.fastmmw.com/2016/05/31/fiji-produo-vs-hawaii-295x2-clblas-preliminary-performance/

**Comment by kknox** [comment]
I tried compiling your package today; I am attempting to reproduce with the /develop branch.  It has a funny `#include <header.h>` that you are not providing or generating.  I renamed it as `<clAmdBlas.h>` as that seemed to be missing too.  

I don't have the definition for START_CLOCK nor END_CLOCK.

Also, give me the exact command to reproduce the kernel fail; right now I'm going to assume
`./Fiji-sgemm`.

Does it fail with a single gpu system, or do I have to have multiple GPU's installed?

Kent

**Comment by paolodalberto** [comment]
Sorry without it it is impossible 
[header.h.zip](https://github.com/clMathLibraries/clBLAS/files/298527/header.h.zip)

You will need two ... 
thank you

**Comment by kknox** [comment]
Hi @paolodalberto 
I have an opencl machine now set up  with 2 Fiji's.  I can compile your sample code. Here is the output I see:
```bash
kknox@puter:~/src/clblas-bug
$ ./Fiji-sgemm 0 8192 8192 8192 2
Start 
----------> get time 2.220192e+00 sec<------
0 GFLOPS 990.465354 
done
kknox@puter:~/src/clblas-bug
$ ./Fiji-sgemm 0 8192 8192 8192 2
Start 
----------> get time 2.221672e+00 sec<------
0 GFLOPS 989.805541 
done
kknox@puter:~/src/clblas-bug
$ ./Fiji-sgemm 1 8192 8192 8192 2
Start 
----------> get time 2.211994e+00 sec<------
1 GFLOPS 994.136176 
done
kknox@puter:~/src/clblas-bug
$ ./Fiji-sgemm 1 8192 8192 8192 2
Start 
----------> get time 2.202018e+00 sec<------
1 GFLOPS 998.640000 
done
```
How do I reproduce your issue?  Your test.sh file is hard-coded to assume 4 installed devices, is there a way to reproduce with 2 Fiji's?

**Comment by paolodalberto** [comment]
did you try with 8000 ? 
the dimension makes difference ... there is a gap for me within a specific interval. 
During the weekend I will git the PCI extension and I will build the system again .... so I will try again

**Comment by kknox** [comment]
I did try 8000, i just mistakenly cut-n-paste the 8192 run into the message above.  i would like to know if you get a segmentation fault if you use the same (fiji/fiji) vs when you use mixed devices (fiji/hawaii).

**Comment by paolodalberto** [comment]
hmmm, with develop-branch  ?
I did have problems with all combinations .... even with release code

**Comment by paolodalberto** [comment]
My opencl compiler ?

**Comment by paolodalberto** [comment]
Looking at my results: 295x2 did not have any issues using Release code
ProDuo did

**Comment by kknox** [comment]
Yes, I'm using the develop branch.  I have 2 Fiji Nano's installed in my machine, which should emulate your ProDuo.  If you modify your tests to only run using ASICS from 1 family (don't mix Fiji and Hawaii's  in a run), do the segfaults disappear?

**Comment by paolodalberto** [comment]
I used the release distribution for Fiji + ProDuo only, seg fault are there

Then I took the develop branch, but I move as a global variable the hash map used for the code generation and removed the "thread" qualifier because of too high over head ... and because using two queues the code will be compiled twice anyway (but sequentially in a non blocking function) 

0.9 TFLOPS for 8000x8000x8000 using two Fiji is too small, I get about 3TFLOPS

I will come back to you and help to details when seg faults happens using only Fijis

**Comment by paolodalberto** [comment]
@kknox I think the real test for you will be sgemm3 (not sgemm)
check test.sh for a list of examples.

**Comment by paolodalberto** [comment]
To remove the compilation out of the equation, I pre-run the GEMM as warm up sequentially and waiting for the results. This trick compiles the sources and creates binary kernels that will be simply recalled later using the key map structure. So far so good: no need of funky synchronizations. 

At the end of my multi-graphic code, I wait for all events associated to write back into memory the results. The segfault happen at this stage. This is from develop branch where I generate code for Fiji and for Hawaii. For both architectures, I can see failures but for different sizes. 

Is there a default and safe matrix matrix kernel that I could call instead ? so to switch the computation path ?

**Comment by kknox** [comment]
Is the segfault happening in a machine with hybrid graphics? (you are using both fiji and hawaii in the same program)
If so, have you tried to run multi-gpu with only 1 product family?  (only fiji's or only hawaii's)

**Comment by paolodalberto** [comment]
I change the setting of my configuration so that I can see 5 GPUs on a crossblade ranger
This is run using develop-branch for Hawaii architecture

0 turk (not used on slot 0) 
1 Fiji 
2 Fiji   1-2 Fiji is a ProDuo on a slot 3  (half the communication speed) 
3 Hawaii
4 Hawaii,  3-4 295x2 (slot 1) 

I change the code in order to balance the computation: Fiji's will use half of the data of Hawaii to compensate for the half the communication speed

This is a successful experiment   
6000
1 GFLOPS 980.422778              // sgemm GPU 1 Fiji 
3 GFLOPS 1995.322654            // sgemm GPU 3 Hawaii 
N 1-3 GFLOPS 816.940683       // Only ProDuo Fiji
N 2-4 GFLOPS 1653.111566     // one FiJi and One Hawaii
N 3-5 GFLOPS 1870.150312     //  ony 295x2  Hawaii 
N 1-5 GFLOPS 1166.471478     //  all ProDuo all 295x2
N 1-4 GFLOPS 1118.339943     //  ProDuo Fiji + 1 Hawaii  
N 2-5 GFLOPS 2032.498690     //  1 Fiji + 295x2


N 7000
1 GFLOPS 1117.104715            // sgemm GPU 1 Fiji
3 GFLOPS 2211.867305            // sgemm GPU 3 Hawaii 
N 1-3 GFLOPS 1005.852898     // sgemm3 only ProDuo
FAILED                                    // one FiJi and One Hawaii
N 3-5 GFLOPS 2452.677799     // sgemm3 only 295x2
N 1-5 GFLOPS 1540.064506     //  all ProDuo all 295x2
N 1-4 GFLOPS 1383.164942     //  ProDuo Fiji + 1 Hawaii
N 2-5 GFLOPS 2376.769812     //  1 Fiji + 295x2

8000
1 GFLOPS 1233.280873            // sgemm GPU 1 Fiji
3 GFLOPS 2363.502535            // sgemm GPU 3 Hawaii
FAILED                                    // sgemm3 only ProDuo
N 2-4 GFLOPS 2337.891537      // one FiJi and One Hawaii
FAILED                                      // sgemm3 only 295x2
N 1-5 GFLOPS 1735.691110       //  all ProDuo all 295x2
N 2-5 GFLOPS 2673.506823       //  1 Fiji + 295x2

Of course this configuration is useless: the most expensive card cannot be fed half speed and thus today I will be changing the configuration using only slot 0 and 1. 

The warm-ups succeed, and these are just calling the kernels and wait to finish. The one failing are non blocking and do not wait for the kernel, just chaining the events of the kernel to write back to memory. Then I wait for that memory write to finish. So, I am sceptical because the size are fine and it will fail more often due to the simplicity of the code;  but still possible. If there was a "safe kernel" without optimization I would use it for this problem size.  The sizes above do not request for "special" size.

A note: Looking at the code of  
template<typename Precision>
clblasStatus
clblasGemm(
 
1) There is a dangerous use of the final events: the code requires a single output event and Gemm it may require multiple kernels to be put in the queue(s). If the Queue is out-of-order execution, the single event is funky, unfortunately is the user that can set up the  queue. IT will be better to have multiple events as output and pass as output how many kernels have been enqueued.  This unfortunately change the interface. 

2) Multiple Queue are welcome: but only Queue[0] is used for the kernel generation ... this could be funy. 

I can see that OpenCL non-blocking execution uses threads ... I will try more configuration and computing styles to see if I can isolate the problem to the kernel execution 

thank you
